### DISCUSSION TITLE
Why so many body types?

### DISCUSSION THREAD

**Discussion by khvzak** [ntex-rs/ntex#192] [Q&A] [answered]
Hello,
Out of curiosity, why ntex (+actix-web) has many different body types?
Eg:
`http::body::Body`
`http::client::MessageBody`
`http::Payload`

maybe the same question for request/response types:
`http::Request`
`http::client::ClientRequest`
`web::WebRequest`

`http::Response`
`web::WebResponse`
`http::client::ClientResponse`

Just to compare, axum/hyper uses a single type, from the `http` crate.

Thanks!

**Comment by fakeshadow** [comment]
because real world http types are complicated. In order to express them you either have a single type with compromise or use specialized types for low overhead.

let's take a very common use case for example:
```rust
struct Reject<S> {
    service: S,
}

// ignore the unconstraint B type.
impl<S,  B> Service<Request> for Reject<S>  
where
    S: Service<Request, Response = Response<B>>
{
    async fn call(&self, req: Request) -> Result<Response<B>, Error> {
        if req.path != "/" {
            return Ok(Response::new(<what_type_is_this>))
        }  
        self.service.call(req).await
    }
}
``` 
in ntex you can express it with `Body` type as it has explicit variants for Bytes. You can also nest Body types with the help of `MessageBody` trait object variant if you want to.
but in axum you would likely be told to use `Box<dyn HttpBody>` for the Bytes which is an additional heap allocation.

**Comment by fakeshadow** [comment, marked as answer]
As for `Payload` type the benefit of separating request/response body type is mainly you can relax the `Unpin` bound. From server's view request body type can be always `Unpin` while response body type can be `!Unpin` for special body use. This is not something often used but it leaves the possibility open. From client's view this still holds in reverse order.

**Reply by khvzak** [reply]
Thank you!
